processing/src/main/java/org/apache/druid/query/groupby/GroupByStatsProvider.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.groupby;
+
+import org.apache.druid.guice.LazySingleton;
+import org.apache.druid.query.QueryResourceId;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Metrics collector for groupBy queries like spilled bytes, merge buffer acquistion time, dictionary size.
+ */
+@LazySingleton
+public class GroupByStatsProvider
+{
+  private final Map<QueryResourceId, PerQueryStats> perQueryStats;
+  private final AggregateStats aggregateStatsContainer;
+
+  public GroupByStatsProvider()
+  {
+    this.perQueryStats = new ConcurrentHashMap<>();
+    this.aggregateStatsContainer = new AggregateStats();
+  }
+
+  public PerQueryStats getPerQueryStatsContainer(QueryResourceId resourceId)
+  {
+    if (resourceId == null) {
+      return null;
+    }
+    return perQueryStats.computeIfAbsent(resourceId, value -> new PerQueryStats());
+  }
+
+  public synchronized void closeQuery(QueryResourceId resourceId)
+  {
+    if (resourceId == null || !perQueryStats.containsKey(resourceId)) {
+      return;
+    }
+    PerQueryStats container = perQueryStats.remove(resourceId);
+    aggregateStatsContainer.addQueryStats(container);
+  }
```

Comment:
This is not intuitive. The callers won't remember calling `closeQuery()`. Is it possible to mold it as a Closer/Closable? PerQueryStats can extend Closeable, and the close can do that. 

It is still not good since it doesn't seem cool to close "Stats" in first place.  Perhaps renaming `PerQueryStats` to `PerQueryStatsContainer` would also help.

Reply:
>  it would require to be explicitly registered with the query closer

Yes. Implementing the interface makes it easier for the developer to call the closer. I have had experience in the past where an implicit contract was not apparent which led to resource leak